OutOfSchool/OutOfSchool.WebApi/Services/Images/ImageDependentEntityImagesInteractionMediator.cs

```diff
@@ -21,152 +21,168 @@ namespace OutOfSchool.WebApi.Services.Images
-        /// <param name="repository">Repository with images.</param>
         /// <param name="limits">Describes limits of images for entities.</param>
         /// <param name="logger">Logger.</param>
-        public ImageInteractionBaseService(
+        public ImageDependentEntityImagesInteractionMediator(
             IImageService imageService,
-            IEntityRepositoryBase<TKey, TEntity> repository,
-            ImagesLimits<TEntity> limits,
-            ILogger<ImageInteractionBaseService<TEntity, TKey>> logger)
+            IOptions<ImagesLimits<TEntity>> limits,
+            ILogger<ImageDependentEntityImagesInteractionMediator<TEntity>> logger)
         {
             ImageService = imageService;
-            Repository = repository;
-            Limits = limits;
+            Limits = limits.Value;
             Logger = logger;
         }
 
         /// <summary>
         /// Gets a logger.
         /// </summary>
-        protected ILogger<ImageInteractionBaseService<TEntity, TKey>> Logger { get; }
+        protected ILogger<ImageDependentEntityImagesInteractionMediator<TEntity>> Logger { get; }
 
         /// <summary>
         /// Gets a service for interacting with an image storage..
         /// </summary>
         protected IImageService ImageService { get; }
 
-        /// <summary>
-        /// Gets a repository with images.
-        /// </summary>
-        protected IEntityRepositoryBase<TKey, TEntity> Repository { get; }
-
         /// <summary>
         /// Gets limits of images for entity of <c>TEntity</c> type.
         /// </summary>
         protected ImagesLimits<TEntity> Limits { get; }
 
         /// <inheritdoc/>
-        public async Task<Result<string>> UploadImageAsync(TKey entityId, IFormFile image)
+        public async Task<Result<string>> AddImageAsync(TEntity entity, IFormFile image)
+            => await UploadImageProcessAsync(entity, image).ConfigureAwait(false);
+
+        /// <inheritdoc/>
+        public async Task<OperationResult> RemoveImageAsync(TEntity entity, string imageId)
+            => await RemoveImageProcessAsync(entity, imageId).ConfigureAwait(false);
+
+        public async Task<ImageChangingResult> ChangeImageAsync(TEntity entity, string oldImageId, IFormFile newImage)
         {
-            if (image == null)
+            _ = entity ?? throw new ArgumentNullException(nameof(entity));
+            Logger.LogTrace("Updating an image the for entity was started");
+
+            var result = new ImageChangingResult();
+
+            if (!string.IsNullOrEmpty(oldImageId))
             {
-                return Result<string>.Failed(ImagesOperationErrorCode.NoGivenImagesError.GetOperationError());
+                result.RemovingResult = await RemoveImageProcessAsync(entity, oldImageId).ConfigureAwait(false);
+                if (!result.RemovingResult.Succeeded)
+                {
+                    return result;
```

Comment:
Shouldn't another `Log.Trace` be called here?

Reply:
See similar pattern in next functions.

Probably we need more error-prone approach ,and from my experience I see two and half possible ways to avoid such problem of missing `Log.Trace()` near all returns: 

1. either use explicitly `try-finally` block wrapping whole method body like:
```
public async Task<ImageChangingResult> ChangeImageAsync(TEntity entity, string oldImageId, IFormFile newImage)
{
    _ = entity ?? throw new ArgumentNullException(nameof(entity));
    try 
    {
        Logger.LogTrace("Updating an image the for entity was started");
        //... method code with return operators in different places
    }
    finally
    {
        Logger.LogTrace("Updating an image the for entity was finished");
    }
}
```

1,5. same as previous but implicit `try-finally` created by `using` block with special disposable class (like mentioned below `MethodTraceLog` that might log "start" on constructing and "finish" on `Dispose`) - bit worse from CPU and used memory but neater in code, esp. when you can use new `using` syntax (without block {}):
```
public async Task<ImageChangingResult> ChangeImageAsync(TEntity entity, string oldImageId, IFormFile newImage)
{
    _ = entity ?? throw new ArgumentNullException(nameof(entity));
    using(var traceLog = new MethodTraceLog("Updating an image the for entity"))
    {
        //... method code with return operators in different places
    }
}
```

2. or using some post processors (PostSharp or any other)

Calling @DmyMi to thread, probably he can add more reasons pro or contra